language-plutus-core/src/Language/PlutusCore/Check/Normal.hs

```diff
@@ -2,105 +2,56 @@
-termValue t                      = builtinValue t
+check :: Term tyname name a -> Either (NormalizationError tyname name a) (Term tyname name a)
+check = transformMOf termSubterms (termSubtypes normalType)
 
 {- Note [Builtin applications and values]
-An older version of the specification had a special case for builtin applications being
-term values. This is important, because they obviously can't be reduced before runtime.
-However, it was missing a corresponding case for builtin *type* applications, which
-resulted in types like `[(con integer) (con 64)]` not being considered normalized, which
-effectively prevents you using integers anywhere.
-
-The current version of the specification has moved to fully saturated builtins and builtin types,
-but the implementation is not there. Consequently we include the special cases for builtin applications
-and also consider builtin types and type level integers to be neutral types.
+An older version of the specification had a special case for builtin type applications being
+normal types. This is important, because they obviously can't be reduced before runtime.
+This resulted in types like `[(con integer) (con 64)]` not being considered normalized, which
+effectively prevents you using integers anywhere (note: this isn't so much of a problem now
+integers aren't parameterized, but it's still wrong).
+
+The current version of the specification has moved to fully saturated builtins,
+but the implementation is not there. Consequently we consider builtin types to be neutral types.
 -}
 
--- See note [Builtin applications and values]
-builtinValue :: Term tyname name a -> Either (NormalizationError tyname name a) (Term tyname name a)
-builtinValue t@Constant{}    = pure t
-builtinValue (TyInst l t ty) = TyInst l <$> builtinValue t <*> pure ty
-builtinValue (Apply l t t')  = Apply l <$> builtinValue t <*> termValue t'
-builtinValue t               = Left $ BadTerm (termLoc t) t "builtin value"
-
-isTypeValue :: Type tyname a -> Bool
-isTypeValue = isRight . typeValue
-
--- ensure that a type is a type value
-typeValue :: Type tyname a -> Either (NormalizationError tyname name a) (Type tyname a)
-typeValue = cataM aM where
-
-    aM ty | isTyValue ty = pure (embed ty)
-          | otherwise    = neutralType (embed ty)
+isNormalType :: Type tyname a -> Bool
+isNormalType = isRight . normalType
 
-    isTyValue TyFunF{}     = True
-    isTyValue TyForallF{}  = True
-    isTyValue TyIFixF{}    = True
-    isTyValue TyLamF{}     = True
-    isTyValue TyBuiltinF{} = True
-    isTyValue _            = False
+normalType :: Type tyname a -> Either (NormalizationError tyname name a) (Type tyname a)
+normalType (TyFun l i o)        = TyFun l <$> normalType i <*> normalType o
+normalType (TyForall l tn k ty) = TyForall l tn k <$> normalType ty
+normalType (TyIFix l pat arg)   = TyIFix l <$> normalType pat <*> normalType arg
+normalType (TyLam l tn k ty)    = TyLam l tn k <$> normalType ty
```

Comment:
What about `TyCon`? That should be a type value, no?

Reply:
`TyCon` isn't a thing - type constants in the spec are represented as `TyBuiltin`, which is in the netural types case. (Why is it `TyBuiltin`? I have no idea, it's always been like that so  maybe you remember?)

Reply:
Right. Should we not update the neutral types case to follow the spec then? And put `TyBuiltin` here?

Reply:
The spec is wrong, builtin types should be neutral, see the note. `[(con someType) someOtherType]` is certainly normal, since builtin types don't have reduction behaviour. I'm going to fix this once the whole normal/value discussion is sorted.

There is no such thing as "builtin types" in the spec at the moment. I think possibly in the past it had both "builtin types" and "type constants", now it seems that the spec just has "type constants" and the implementation just has "builtin types", but they're the same. We should just rename them to be consistent. I think that "builtin types" is probably right: they're more like term builtins in that they're named.

Reply:
Ah okay, sounds good. Thanks.